### PR TITLE
Fix i3dm matrices in ModelExperimental

### PR DESCRIPTION
**Pull request by ptrgags**
Fixes https://github.com/CesiumGS/cesium/issues/10255

This PR fixes two things about i3dm model matrices in `ModelExperimental`:

1. The matrix product was missing `model.modelMatrix`
2. The axis correction was applied on the right of the node hierarchy transform, when it should have been applied on the left.

I also refactored how axis correction is handled, now it's just a matrix `sceneGraph.axisCorrectionMatrix`. 

Here's a diagram that shows the correct matrix product order and how it compares with the glTF `EXT_mesh_gpu_instancing` matrix product.

image

Sandcastles:

[sandcastle from issue](http://localhost:8080/Apps/Sandcastle/index.html#c=7Vl7b9s4Ev8qPB+wta8O47y6PTvpXZq4aIDtpte4LRZ1saAlxuZGFnUU5cRb+Lvf8CWRshw77f55QZtI5HCeP84MqQuas2KOhw8ZFWxOU0mSN5TIQtAc05RMEvqOxzTx59EZkqKgg3E6Tvf30XkcI4KihGUZS6coS0hKuzCiH9CU8jmVYokkR/mM3yM5o0jQDPgrZpLxFPFbNaqZudVpjCKeSsETvWBOpimTRUwtrWUes1ySNKLonsmZIeRFTrHSDJbnEi0YvacCNE7pPbowtn7SY+1xK9LvFyCHsJSKcauLvo1ThFh6y1/zhz66JUlOu2oopwmNlLJXacwiIrmoZledgROXRxTUOrNisX4dVNooJ11P/gBOORB9GbeuUmNA3EUpB+XB02CrFOxBKePNGwtDgq8D30odJuCqLYjppJi+5gUom04/8aSY03yooxn7VtF4Sm/kMgGiclaF1kw+kHmW0NEyo3nfV13PRoUQEMBhRRTQfOl9VZ7RtidUIknElMrfQL8e7g3MmI7hMJVMMqr98dVOGGfT+L0isGMOX3osd9i70ZQWDfczmhoAoJjfl05Xz28BUEkdBzeRoDS9yUhEhwuwxRK1lXl+AHFE0gXJx6mKs8cN51RepVkhzzUy2rdFqh9Qe84XVO2VjomG0SNj0R2NjX9AEcNaDZbkOOM5UxyUIIDhLdK6IKdwTG8Bp3Hb59RBP/20lQizeFc6rH3ZUaRWexTGA1RvXDJooMUAVUgbOmVYsRc84QJ/fns1GmKF6fMkm5E2gOKk08iBFxLgSfWyRi5ulXZnXoX0gsypIBcmiahomWymA6bQpneBXrwCoHY3gUIhG/8yfDP6/fL6868dB7wPoCTJXR7iDndF5lBXZH8d5kpemxFnY6UgU4tv4M5Oc0yfEKeDHwxTDUqQoYyeLhI62T41lrYc7RLFj+/LGH7MYjC7HkK1GV0Q1fNfF0aP286pY/eA2mRHE0lUmi1zChQQId/bxIKXaK+aomlcTdgAuUT9/Mzy2nGLvLv+eDP8/d31p6Hxb2kREIOXtbLG4W/sTNtkGqO/oNBypGgN0MiEB5eV/sxpaPW1C6sUpGrOqiw7LIE+Qw48fRJO4pEZbhcicSk6KC9hsC/8uQsFQKO/ryBUyC/mFW1c2nYEIQmEB55IeqSyYBfpX3sHuNfpVvQw6F7c8Ff7V9XwzyyWs37VBuD1wo7+hYAn6mv+2k8mTsj5qNa14AzaPYDGAvpAEsdWd19v/efo0vnym1MQnNpXv0r1Q9/2a++WbGVKTqgT1n3MDbSNYS+DzsyiyuBHGp6Bhy/HFzAZL98LPmc5NawwdI8NCbXaWRPL/CabUaFxaHmFE4NwlSAxKxSeQipsxq21zhTw/Z+cz0fcQ4qV0m3EzlswQzmSyWj2gaRTChA6AfT08GHXif4HOgYslejxRFb9hfr5m+X5TneXx5j+t4AK6RFUURGcSywFSfNbLuZdn6TG5Opy+OvoavRbRVJpUjkYIcjGI+jew0MEYjm04oBBkiRL5HojgDKRute36jzLkVIIlQrhgO95lsFqkiJAsWZAXAetjiQAdah8ml1NOIBAjbq4odzE3SzwRJg4l8IvDMOzuicgZY0UTaKPPE91a3O+6HhuHWzWCBbwqSDZjEVeafZG8S3shJJvqFto12MCQ4T/mNTGrYWjrTrMKJvOpEsP5mezXtiS7zVbvE7pywuSGNYnMxPr9dgrQzcFf1slMIIbI72qtjLoi9qq4DF1whrAn9O6hglNp3I2QM+fs2Dr2cOJ7cdqiwC1bbbu5urktqzKBrUnOV0uvvk2uq3bR40BDZCuWfdRsB7Os+rqIQcWUD6aPHbYDulRlf0OISM+NhnOdWq0rjHub++LayutIYG2STIh0R3UnYwKuVzTeVurtGaHPsM8boBtzL1z/dqcNqnJwIB85b2t/DKCwoM8zop81vYQEgDW+j+oxgO/A4BKHBEoZ14ppkJwUavHPKE44VM7V3JwvV/Q7ekO4f+9nncTYfdjmJeNYjpZXdIptPiu+O8dHB7h3s/Hxy8O/mkVOYae4qR39HPvhR046lm1TSxcNtbtSZAQZ1jyDzAM2bB9cHQCpg/8KxKIfHlHVPZQEIa1wVlWO5IFvRCsaFvpXcO1q9n4wrhg5Q1kqeHIZf8cz2oM/1OoXJBWCdy50boANKo7gO6QIlM42/r9cpUsa/w9ffv+i53WFchLnIVgYRueRyQBSS/LgTk0V/Ni/p490OSG/QlzBz2HmN379hBg1lAoouqKqCwRxhGW5juL1e6F6slFaqPPm+rR9lqkN0PX7Al/Z39fMXliIXlSEakVEH9yU+HYuWisul5W9tLnlkKxsincBJG5C/CPIjFt3biF8T78u9EXz5dEkv3gKLpf3plXT9fVbtl3veUfOU/HLXNB7V0J+ALLm6KRgrP5KKG9a04GTLcyucazGlAQM9vQaS85h0CpRBXzqNB3LgDUYaKvX14vr+BY3bI045aSZt15l/LoDhxt9lG7zPVNJESdb15bXSrSrhNujKitUo5UqlxPcioWqmj4K8et9dt90E8vyotJHgk2oV6BBmh+IklBy60qtCuruKuTpiNCZ2dntS8FXm1/LBQaTIgCYnen10dAd8mlo1RSlRcfZMIXEFJoJQgruxoHk7Cx9zbh9ua+ufAfnOiif/ISv+iikx5+eXgSnM7LRmn1XZF75CLksRAugvg13Dpaf3S8WG29pEGa64/btN7h7G6KLcSucXJK+S1SCdNaTfEXlc8Dz4xSpsW7kVmvNIKqC9dz6EhMeDfdshm60tGatvETGVr7muZf4vk36jqRtrqt0xy8R1+5wP2bzTMupOoO2pBJJYVNDlkt358UULUljvK82j9/d0nMK4Gq+EyFinYfiemEtI8Pu8j9B0j7J8ZM3buk0z46zh684QkXMRV75iQWTq7WRDN1Ve4rsICyx6Cf2SMJm0LZnrM4Tui61D3Jsz46DCS7qQmXks/D2XXRug2kgfW3PJV79/pcrk60SVxffrrv+/s0ZgvE4rOG782AOJLnMHNbJLoFG7dene4D/dpSledA6WswPCFLRTY7ePWLGcQYn+7Da/PKsr6U4T81XxAQNAdkT1UxoOKZNB2N/9W3a/ZKHzWVAxBn+FR8TZgkTCtjZzS6m/AHsFJzUVbqbqMVCNZk6rvzI4nL6mHamT56puU88yxSyae6qVvoxTtrpfqbTUqtZ56tuhg6vRLlZqlDhQmPefsf)
[CDB San Diego bridge and trees](http://localhost:8080/Apps/Sandcastle/index.html#c=jVZZc9pIEP4rU9Q+iCosdCEJ4zhr42OhDE5icjmkNkIazBhpxM6McHCK/76tGQkk2zkeEFJP99fX1y31MSdZop9/X2FGEkxFEF/gQGQMcx3TYBbjURrhuHqOXiHBMtyb0ikNU8oFivAcMxD3FZh6fP26fH5YYKqEvdJiTfCDNKH4oVT7IGXatBHK534KzgjFbNpoSl/tNrpkARWF/vAGBSGociRStEkzhkhKUcA5FnxKC8xBKj0HWSxOpPIkXWIKfqcNvBkuZpchuSbDwfvHgTkmAz6g7zphf+AOlqtPH/rDrg5K/0WXS1AaPI6Ti8Xo8TYZ3ZhkdL/cXE3eOqOzkbj+eP543TeXt8nF/ejj5++3yfvN+OOAXPWHq1sAG5+95YMkXkRwP5p8fhhPTszR5KQztg098j4l49nFIHlzbp92yN3w0/C9d+ufBaeGtzrl9+9mS/MfISZn2XLaUBWPIUU0gV4w9GNKEZIFZVkoUqbFwQzHTSVHSCwI16UIEpb/vcoBFwETOQ4cZhSKBLWOqgqYRi8eb/NL/pMIWt1dFRUoM09ZEtAQ6zR90JpPzdPVE+u9y9/YrhihFdcFDckcOIbB5F+OQwDRaqAHTyJsojYyDcPQjd4eJY2xHqd3Wh1r5327J73IW8DBy49tZRRWLE0Ix7n8y1cpn2c0FDk1c4OrNIhUl1ooY2WrlCnDQbRRjVVzIe+1b3/9kAZbpEmNN8pD85sKStkSSgQJ4l9a73RizPM4cFTHSNeYBfGvMQqd0nBvHKU0b5uc8qJZZSn0VcYXWq6gF6K9cdWnXvBJHu6LUROXpYcUsKivD/Vnn03UoVZQA6ose6fsC0s9Cb6TJEtuQoYxvVkFIT5nLM3TdqSe2k86D7GMmiRQujWkEkSRVmDskyhBq+3RBSw9DRj66rgkaS2lnPoF7aq9qyW73buounnexzys8zUs5yvCBUTM6o6fOKi4flL+yolsF7wE0niNn4ejuP9FsuJrPgKVDA8ribZqARzWImnVQjisxSLd9Ypxa7d3kwNbO8brIB+naaOF9kv+HUSaMdgVcyj/Sf4OGESa6budrtHMw65h3HH5TuN/BOHZviMhqgBCAZwyEt3hP4KxLfcZjADy8T+17jTVW7DkF5RKKyesmdfrBcbBBkWaGpkvxdIpSCSrzL+idI6uZ/c4FLAlBSOYa6q1zd1qra9F9YLplUc1NMV/vVjNP9EpGPAbrYILOy2k1LY7Gk5pMaFhAAaBPo83k1SNfIS5IFRyBB3WFgRMFtwF1NYOLKdjdV1Td13LNh3HMdwWOnA8r2uapm67jtO1vY7TQrZjwb+pe5ZjuWbX9t2m5G0KtYLPk8LJrlQRYTj8pWdDd4yu2TEdw7S7jtm1DPByAFLT6fiua9qO7Zlw7Uipb1q+Y+UHruN1Hb/ZKj1lq58nZ+iWb9iu5bm22wECu6by0TUc3wEg2/Msy3f9FjJ02zJsv+vZhu/btm0VHrbyGmWszNCAWWz2Gq3GERebGB+XYfxNklXKRL5lNV1vC5ys4gACac+ycAm7KuS87PFRu2p6FJE1ItGrF772kPzEgZN5Fsc35BEG7PioDfrPTGOYI0LvroEtcbDJ1Rbm8ZUS6rp+1IbHly1FmsazgD1B/h8)

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/i3dm-matrix-fixes/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

**Comment by ptrgags** [comment]
@j9liu this is now ready for review when you get a chance.

File: Source/Scene/ModelExperimental/InstancingPipelineStage.js [hunk 1]
```diff
@@ -142,16 +143,47 @@ InstancingPipelineStage.process = function (renderResources, node, frameState) {
...
+        modelViewScratch
+      );
+
+      // modifiedModelView = view * modifiedModel
+      return Matrix4.multiplyTransformation(
+        frameState.context.uniformState.view,
+        modifiedModelMatrix,
+        modelViewScratch
       );
     };
```
Comment (j9liu) [inline Source/Scene/ModelExperimental/InstancingPipelineStage.js:174]: nitpick, can we add a line break here for readability?

File: Specs/Scene/ModelExperimental/InstancingPipelineStageSpec.js [hunk 2]
```diff
@@ -407,6 +414,7 @@ describe("Scene/ModelExperimental/InstancingPipelineStage", function () {
       const shaderBuilder = renderResources.shaderBuilder;
       const uniformMap = renderResources.uniformMap;
       const runtimeNode = renderResources.runtimeNode;
+      renderResources.model.sceneGraph.components = components;
```
Comment (j9liu) [inline Specs/Scene/ModelExperimental/InstancingPipelineStageSpec.js:417]: I'm not familiar with the i3dm loader so I'm just curious, why is this line necessary?

File: Specs/Scene/ModelExperimental/InstancingPipelineStageSpec.js [hunk 3]
```diff
@@ -427,10 +435,44 @@ describe("Scene/ModelExperimental/InstancingPipelineStage", function () {
         _shadersLegacyInstancingStageVS,
       ]);
 
-      expect(uniformMap.u_instance_modifiedModelView()).toEqual(Matrix4.ZERO);
+      const model = renderResources.model;
+      const sceneGraph = model.sceneGraph;
+
+      // For i3dm, the model view matrix chain has to be broken so the shader
```
Comment (j9liu) [inline Specs/Scene/ModelExperimental/InstancingPipelineStageSpec.js:441]:
```suggestion
      // For i3dm, the model view matrix chain has to be broken up so the shader
```

**Review by j9liu — COMMENTED** [review]
@ptrgags - alright, I tested the Sandcastles and the basic instanced 3D models, and they look good! I think the code overall is fine too. Most of my comments are just to tweak the documentation for clarity.

File: Source/Scene/ModelExperimental/InstancingPipelineStage.js [hunk 4]
```diff
@@ -142,16 +143,47 @@ InstancingPipelineStage.process = function (renderResources, node, frameState) {
...
+    // LegacyInstancingStageVS.glsl. Conceptually the product looks like this:
+    //
+    // modelView = u_modifiedModelView * a_instanceTransform * u_nodeTransform
     uniformMap.u_instance_modifiedModelView = function () {
-      return Matrix4.multiply(
-        frameState.context.uniformState.view,
+      // Model matrix without the node hierarchy or axis correction
+      // (see u_instance_nodeTransform).
+      const modifiedModelMatrix = Matrix4.multiplyTransformation(
+        // Since this is 3D Tiles, model.modelMatrix is the computed tile
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Scene/ModelExperimental/InstancingPipelineStage.js:159]: yes though this legacy path only gets applied for i3dm which is only in 3D tiles. I'll clarify.

File: Specs/Scene/ModelExperimental/InstancingPipelineStageSpec.js (hunk 2 above)
Comment (ptrgags) [reply Specs/Scene/ModelExperimental/InstancingPipelineStageSpec.js:417]: oh it's that the render resources are being mocked here, so we need to insert the loaded components. I'll leave a comment explaining this.

**Comment by ptrgags** [comment]
@j9liu updated!

**Comment by j9liu** [comment]
@ptrgags looks good, merging now. Thanks!
